src/IceRpc.Slice.Generators/Internal/Parser.cs

```diff
@@ -0,0 +1,202 @@
+                        parentType.Keyword.ValueText);
+                    container = container.Parent;
+                    parentNode = parentNode.Parent;
+                }
+            }
+        }
+        return serviceDefinitions;
+    }
+
+    private AttributeData? GetAttribute(ISymbol symbol, INamedTypeSymbol attributeSymbol)
+    {
+        ImmutableArray<AttributeData> attributes = symbol.GetAttributes();
+        foreach (AttributeData attribute in attributes)
+        {
+            if (SymbolEqualityComparer.Default.Equals(attribute.AttributeClass, attributeSymbol))
+            {
+                return attribute;
+            }
+        }
+        return null;
+    }
+
+    private string GetFullName(ISymbol symbol)
+    {
+        if (symbol is INamespaceSymbol namespaceSymbol && namespaceSymbol.IsGlobalNamespace)
+        {
+            return "";
+        }
+        else
+        {
+            string containingSymbolName = GetFullName(symbol.ContainingSymbol);
+            return containingSymbolName.Length == 0 ? symbol.Name : $"{containingSymbolName}.{symbol.Name}";
+        }
+    }
+
+    private IReadOnlyList<ServiceMethod> GetServiceMethods(ImmutableArray<INamedTypeSymbol> allInterfaces)
+    {
+        Debug.Assert(_operationAttribute is not null);
+        var allServiceMethods = new List<ServiceMethod>();
+        foreach (INamedTypeSymbol interfaceSymbol in allInterfaces)
+        {
+            allServiceMethods.AddRange(GetServiceMethods(interfaceSymbol));
+        }
+        return allServiceMethods;
+    }
+
+    private IReadOnlyList<ServiceMethod> GetServiceMethods(INamedTypeSymbol interfaceSymbol)
+    {
+        var serviceMethods = new List<ServiceMethod>();
+        foreach (IMethodSymbol method in interfaceSymbol.GetMembers().OfType<IMethodSymbol>())
+        {
+            if (GetAttribute(method, _operationAttribute!) is not AttributeData attribute)
+            {
+                continue;
+            }
+
+            foreach (TypedConstant typedConstant in attribute.ConstructorArguments)
+            {
+                if (typedConstant.Kind == TypedConstantKind.Error)
+                {
+                    // if a compilation error was found, no need to keep evaluating other args
+                    return serviceMethods;
+                }
+            }
+
+            ImmutableArray<TypedConstant> items = attribute.ConstructorArguments;
+            Debug.Assert(
+                items.Length == 1,
+                "Unexpected number of arguments in attribute constructor.");
+            string operationName = (string)items[0].Value!;
+            serviceMethods.Add(new ServiceMethod
+            {
+                OperationName = operationName,
+                DispatchMethodName = GetFullName(method)
+            });
+        }
+        return serviceMethods;
+    }
+
+    private bool HasServiceAttribute(INamedTypeSymbol classSymbol)
```

Comment:
I was confused by this code (see below). It would be good to rename classSymbol to baseClassSymbol and add a comment like:
```csharp
<summary>Checks whether this base class or any of its base class(es) has the SliceService attribute.
When true, we'll override the DispatchAsync method generated for this base class; otherwise, we'll implement the IDispatcher interface.</summary>
```

Reply:
Maybe you should replace this method by:
```csharp
/// <summary>Returns the nearest base class with the SliceService attribute.</summary>
/// <value>The nearest base class with the SliceService attribute. null when there is no such base class.</value>
private INamedTypeSymbol? GetBaseServiceClass(INamedTypeSymbol classSymbol)
```

See other comment about the base class logic.